### ISSUE TITLE
NodesDataView is not always a set, but promises that it is

### ISSUE THREAD

**Issue by NeilGirdhar** [networkx/networkx#7176]
I was just looking at filling in some annotations on the typeshed, and I ran into a consistency issue with some definitions.  `NodesDataView` is only a set when its `data` element is false.  Otherwise, it violates the set interface.  This would be very confusing if it were used as a set in generic code.

I suggest making a copy of this class: `NodesDataViewSet`, which covers the special case when it is a set.  The `NodesDataView` can then make a more conservative promise of being a `Container`.

**Comment by dschult** [comment]
Actually, `NodeDataView` is a subclass of `set` and it is supposed to act like a set when the graph data is hashable. But, as always, floating-point data can be difficult due to round-off errors. Thus the class is perhaps well described as a very finicky set in the doc_string:
```
    The main use for this class is to iterate through node-data pairs.
    The data can be the entire data-dictionary for each node, or it
    can be a specific attribute (with default) for each node.
    Set operations are enabled with NodeDataView, but don't work in
    cases where the data is not hashable. Use with caution.
    Typically, set operations on nodes use NodeView, not NodeDataView.
    That is, they use `G.nodes` instead of `G.nodes(data='foo')`.
```
If your data values are hashable (say integers), doesn't that work?

Can you give an example where there aren't floating point issues or unhashable issues where a problem arises?
This is similar to trying to use a python set on your data. In many cases it won't work for you.

**Comment by NeilGirdhar** [comment]
> Actually, `NodeDataView` is a subclass of `set` and it is supposed to act like a set when the graph data is hashable.
> If your data values are hashable (say integers), doesn't that work?

It works great when `data` is false.  But when it's true, it does not match the set interface:

```python
from networkx import Graph

g = Graph([(1, 2), (2, 3)])

data_is_set = g.nodes.data(False)
data_is_not_set = g.nodes.data(True)

print(list(data_is_set), data_is_set & {2}, 1 in data_is_set)
print(list(data_is_not_set), data_is_not_set & {2}, 1 in data_is_not_set)
```

The `collections.abc.Set` over elements of type `int` defines `__iter__`:
```python
    def __iter__(self) -> Iterator[int]: ...
```
But when `data` is true, the `__iter__` method returns `Iterator[tuple[int, dict]]`

**Comment by dschult** [comment]
Yes -- it looks like the `contains` routine for that class is broken. 
We test contains explicitly for `(node, data) in G.nodes.data(...)` because that is how is it intended.  But apparently only indirectly (in the `test_pickle` test) for `node in G.nodes.data(...)`. That test passes, but for the wrong reasons. 

Checking `node in G.nodes.data('weight')` when `node` is in `G` currently returns `True`.
So I guess this is a set with nodes allowed membership but never reported on the `__iter__` membership list.  That pretty clearly breaks the rules for sets.

I'm pretty sure that is not what was intended, but it was a long time ago and while I wrote it I can't vouch for any specific intentions here. Certainly, that code hasn't been exercised much. Luckily it is very unlikely to have affected users -- they would need to be checking for node membership (which includes equality as sets) in a node-data view. All of our examples and tests check node-data pairs for membership. As the name suggests, it should be all about node-data pairs.

That said, who knows, maybe someone has written their code to create a NodeDataView and then use it to check node membership in a graph. So, while this probably doesn't matter in practice, let's figure out:
- should we return False for `node in G.nodes.data("weight")` when `node` is in `G`?
- or should we raise an exception for `node in G.nodes.data("weight")` when `node` is in `G`?
And do we deprecate this behavior or change it?

I guess I'm in favor of an exception, and an immediate change... but I could be persuaded otherwise.
Switching from returning True to returning False could be surprising to folks who are using its current behavior. While raising as exception with a note that they probably wanted to check membership in the graph, or in a NodeView rather than a NodeDataView would be less silent -- and hopefully clear.

All of this has the usual caveats that we learn to live with in NX: if you have a node (1, 2) and you also have a node 1, you can run into confusing results. In this case, if node `1` has a node attribute with value `2`, the tuple (1, 2) is a node and also a member of a NodeDataView (and also potentially an edge if `1` is connected to `2`).

**Comment by NeilGirdhar** [comment]
I'll answer more in detail later, but I think the design of NetworkX in this regard is pretty beautiful.  Are you opposed to returning different types for the true and false case?  Would you like me to create a pull request so that you can see what that would like?

**Comment by NeilGirdhar** [comment]
> Yes -- it looks like the contains routine for that class is broken.

It's not just contains though: iter would have to be adjusted to match the set interface.  And if you change iter, I imagine that would break a lot of code?

> * should we return False for `node in G.nodes.data("weight")` when `node` is in `G`?
> * or should we raise an exception for `node in G.nodes.data("weight")` when `node` is in `G`?
>   And do we deprecate this behavior or change it?

Would you consider keeping NetworkX's current behavior, which  I think is as useful as possible?  Instead just adjust the interface promises so that:

```python
nodes.data(False)  # inherits from Set[Node]
nodes.data(True)  # inherits from Container[Node], Sized and Iterable[Node, dict[str, Any]]
```
This could be achieved by adding a separate class.

**Comment by dschult** [comment]
I'm not sure we should put much effort into this feature (not sure there is even a use for a set-like object of `(node, data)` pairs).

**Comment by NeilGirdhar** [comment]
Fair enough!  Just instinctively polishing minor things.

> not sure there is even a use for a set-like object of (node, data) pairs).

Oh I agree that there isn't.  I just wasn't sure what you were imagining.
